transport/x/grpc/headers.go

```diff
@@ -22,82 +22,102 @@ package grpc
 
 import (
 	"fmt"
-	"strings"
 
 	"go.uber.org/yarpc/api/transport"
+	"go.uber.org/yarpc/transport/x/grpc/grpcheader"
 
 	"go.uber.org/multierr"
 	"google.golang.org/grpc/metadata"
 )
 
-const (
-	globalHeaderPrefix      = "yarpc-grpc-"
-	reservedHeaderPrefix    = globalHeaderPrefix + "reserved-"
-	applicationHeaderPrefix = globalHeaderPrefix + "app-"
-	callerHeader            = reservedHeaderPrefix + "caller"
-	encodingHeader          = reservedHeaderPrefix + "encoding"
-	serviceHeader           = reservedHeaderPrefix + "service"
-)
+// TODO: there are way too many repeat calls to strings.ToLower
 
 // transportRequestToMetadata will populate all reserved and application headers
 // from the Request into a new MD.
 func transportRequestToMetadata(request *transport.Request) (metadata.MD, error) {
 	md := metadata.New(nil)
-	err := addToMetadata(md, callerHeader, request.Caller)
-	err = multierr.Append(err, addToMetadata(md, encodingHeader, string(request.Encoding)))
-	err = multierr.Append(err, addToMetadata(md, serviceHeader, request.Service))
-	err = multierr.Append(err, addApplicationHeaders(md, request.Headers))
-	return md, err
+	if err := multierr.Combine(
+		addToMetadata(md, grpcheader.CallerHeader, request.Caller),
+		addToMetadata(md, grpcheader.ServiceHeader, request.Service),
+		addToMetadata(md, grpcheader.ShardKeyHeader, request.ShardKey),
+		addToMetadata(md, grpcheader.RoutingKeyHeader, request.RoutingKey),
+		addToMetadata(md, grpcheader.RoutingDelegateHeader, request.RoutingDelegate),
+		addToMetadata(md, grpcheader.EncodingHeader, string(request.Encoding)),
+	); err != nil {
+		return md, err
+	}
+	return md, addApplicationHeaders(md, request.Headers)
 }
 
-// populateTransportRequest will populate the Request with all reserved and application
-// headers from the MD.
-func populateTransportRequest(md metadata.MD, request *transport.Request) error {
-	caller, callerErr := getFromMetadata(md, callerHeader)
-	request.Caller = caller
-	encoding, encodingErr := getFromMetadata(md, encodingHeader)
-	request.Encoding = transport.Encoding(encoding)
-	service, serviceErr := getFromMetadata(md, serviceHeader)
-	request.Service = service
-	headers, headersErr := getApplicationHeaders(md)
-	request.Headers = headers
-	return multierr.Combine(callerErr, encodingErr, serviceErr, headersErr)
+// metadataToTransportRequest will populate the Request with all reserved and application
+// headers into a new Request, only not setting the Body field.
+func metadataToTransportRequest(md metadata.MD) (*transport.Request, error) {
+	request := &transport.Request{
+		Headers: transport.NewHeadersWithCapacity(md.Len()),
+	}
+	for header := range md {
+		header = transport.CanonicalizeHeaderKey(header)
+		value, err := getFromMetadata(md, header)
```

Comment:
Ranging over `md` gives you the values already. Another map lookup on it is
unnecessary.

```go
for header, values := range md {
  var value string
  switch len(values) {
  case 0:
    // leave value empty
  case 1:
    value = values[0]
  default:
    return nil, fmt.Errorf(...)
  }
  
  // ...
}
```